Agents/BMSUpdateAgent/BMSUpdateAgent/src/main/java/uk/ac/cam/cares/jps/agent/bmsupdate/BMSUpdateAgentLauncher.java

```diff
@@ -107,66 +156,163 @@ public JSONObject processRequestParameters(JSONObject requestParams) {
+                rsClient.setPassword(sparqlPassword);
+            }
+            //query for Bacnet Device ID linked to data IRI
+            String deviceId = bmsUpdateAgent.getDeviceId(dataIRI, rsClient);
+            //query for Bacnet Object ID linked to data IRI
+            String objectId = bmsUpdateAgent.getObjectId(dataIRI, rsClient);
+            //send request via API
+            try {
+                BMSWacnetAPIConnector wacnetAPIConnector = new BMSWacnetAPIConnector(System.getenv(ENV_WACNET_API_PROPERTIES));
+                result = wacnetAPIConnector.writePresentValue(deviceId, objectId, value);
+                return result;
+            } catch (Exception e) {
+                throw new JPSRuntimeException("Unable to write present value to Bacnet Object: " + objectId, e);
+            }
+        } else if (requestParams.has(KEY_BACNETDEVICEID) && requestParams.has(KEY_BACNETOBJECTID) && requestParams.has(KEY_VALUE)) {
+            //retrieve Bacnet Device ID from request parameters
+            String deviceId = requestParams.getString(KEY_BACNETDEVICEID);
+            //retrieve Bacnet Object ID from request parameters
+            String objectId = requestParams.getString(KEY_BACNETOBJECTID);
+            Double value = requestParams.getDouble(KEY_VALUE);
+            //send request via API
+            try {
+                BMSWacnetAPIConnector wacnetAPIConnector = new BMSWacnetAPIConnector(System.getenv(ENV_WACNET_API_PROPERTIES));
+                result = wacnetAPIConnector.writePresentValue(deviceId, objectId, value);
+                return result;
+            } catch (IOException e) {
+                throw new JPSRuntimeException("Unable to write present value to Bacnet Object: " + objectId, e);
+            }
+        } else {
+            throw new JPSRuntimeException("missing keys and values in request!");
         }
+    }
 
-        if (!requestParams.has(KEY_CLIENT_PROPERTIES)) {
-            LOGGER.error(KEY_CLIENT_PROPERTIES + " is missing");
-            return false;
+    /**
+     * execute update triples route
+     * @param requestParams Request parameters parsed from the body of POST request
+     * @return results of executed route
+     */
+    public JSONObject executeUpdateTriples(JSONObject requestParams) {
+        JSONObject result = new JSONObject();
+        LOGGER.info("The size is " + requestParams.length());
+        JSONArray checks = requestParams.getJSONArray("checks");
+        for (int i = 0; i < checks.length(); i++){
+            JSONObject checkAndUpdateParameters = checks.getJSONObject(i);
+            String dataIRI = checkAndUpdateParameters.getString(KEY_DATAIRI);
+            String triggerValue = checkAndUpdateParameters.getString(KEY_TRIGGER_VALUE);
+            String insertString = null;
+            String deleteString = null;
+            String clientPropertiesFile = System.getenv(checkAndUpdateParameters.getString(KEY_CLIENT_PROPERTIES));
```

Comment:
Should move line 234-246 out of the for loop. Since the client for each check is the same, will be faster if share the same agent, rsClient and rdbClient instance.

Reply:
For this, the agent is designed to allow the users to input different client.properties files for each check. The users just have to create a properties file with the appropriate keys and values and update the "ENV" section in the stack json file before spinning up the agent.